jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp

```diff
@@ -0,0 +1,272 @@
+(JNIEnv* env, jclass cls, jlong indexPointer, jfloatArray queryVector, jint k)
+{
+	faiss::Index *indexReader = nullptr;
+	try {
+		indexReader = reinterpret_cast<faiss::Index*>(indexPointer);
+		float* rawQueryvector = env->GetFloatArrayElements(queryVector, 0);
+		int dim	= env->GetArrayLength(queryVector);
+
+		std::vector<float> dis(k * dim);
+		std::vector<faiss::Index::idx_t> ids( k * dim);
+		indexReader->search(1, rawQueryvector, k, dis.data(), ids.data());
+		env->ReleaseFloatArrayElements(queryVector, rawQueryvector, 0);
+		has_exception_in_stack(env);
+
+		int resultSize = k;
+		//if result is not enough, padded with -1s
+		for(int i = 0; i < k; i++) {
+			if(ids[i] == -1) {
+				resultSize = i;
+				break;
+			}
+		}
+
+		jclass resultClass = env->FindClass("com/amazon/opendistroforelasticsearch/knn/index/KNNQueryResult");
+		jmethodID allArgs = env->GetMethodID(resultClass, "<init>", "(IF)V");
+		jobjectArray results = env->NewObjectArray(resultSize, resultClass, NULL);
+		for(int i = 0; i < resultSize; ++i) {
+			float distance = dis[i];
+			long  id = ids[i];
+			env->SetObjectArrayElement(results, i, env->NewObject(resultClass, allArgs, id, distance));
+		}
+		has_exception_in_stack(env);
+        return results;
+
+	}
+	catch(...) {
+		if(indexReader) delete indexReader;
+		catch_cpp_exception_and_throw_java(env);	
+	}
+	return NULL;
+}
+
+JNIEXPORT jlong JNICALL Java_com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex_init
+(JNIEnv* env, jclass cls,  jstring indexPath, jobjectArray algoParams, jstring spaceType)
+{
+
+	faiss::Index* indexReader = nullptr;
+	try {
+        const char *indexPathCStr = env->GetStringUTFChars(indexPath, 0);
+        string indexPathString(indexPathCStr);
+        env->ReleaseStringUTFChars(indexPath, indexPathCStr);
+        has_exception_in_stack(env);
+		//whether set IO_FLAGS = 0 or IO_FLAG_READ_ONLYfaiss::IO_FLAG_READ_ONLY
+		indexReader = faiss::read_index(indexPathString.c_str(), faiss::IO_FLAG_READ_ONLY);
+		return (jlong) indexReader;
+	} 
+	catch(...) {
+        if (indexReader) delete indexReader;
+        catch_cpp_exception_and_throw_java(env);	
+	}
+	return NULL;
+}
+
+/**
+ * When autoclose class do close, then delete the pointer
+ * Method GC pointer
+ */
+JNIEXPORT void JNICALL Java_com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex_gc
+(JNIEnv* env, jclass cls,  jlong indexPointer)
+{
+	try {
+		faiss::Index *indexWrapper = reinterpret_cast<faiss::Index*>(indexPointer);
+        has_exception_in_stack(env);
+        delete indexWrapper;
+        has_exception_in_stack(env);
+    }
+    catch (...) {
+        catch_cpp_exception_and_throw_java(env);
+    }
+}
+
+/**
+ * Method: Global Init
+ *
+ */
+JNIEXPORT void JNICALL Java_com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex_initLibrary(JNIEnv *, jclass)
+{
+	//set thread 1 cause ES has Search thread
+	//TODO make it different at search and write
+//	omp_set_num_threads(1);
```

Comment:
Why is this commented out?

Reply:
i am no sure is it appropriate to set threads as 1 or  reuse the settings:
`public static final String KNN_ALGO_PARAM_INDEX_THREAD_QTY = "knn.algo_param.index_thread_qty";`

Reply:
I think it probably makes sense to reuse KNN_ALGO_PARAM_INDEX_THREAD_QTY.